### ISSUE TITLE
[BUG] Tensor Product of observables depends on order of operators and ignores wires information

### ISSUE THREAD

**Issue by PatrickHuembeli** [PennyLaneAI/pennylane#2263]
### Expected behavior

It seems that the tensor product operator `@` only depends on the order of the operators and not on their wire information. Not sure if this is intentional, but seems to be counterintuitive.

Expected behaviour
`(qml.PauliZ(1)@qml.Identity(0)).matrix == (qml.PauliZ(0)@qml.Identity(1)).matrix ` is false

Expected behaviour: Both return statements of the following circuit return the same value
```
dev = qml.device('default.qubit.jax', wires=n_qubits//2)
np.random.seed(1)

@qml.qnode(dev, interface='jax', diff_method="backprop")
def qnode(params):
    for i in range(n_qubits//2):
        qml.RX(jnp.pi*np.random.rand(), wires=i)
    return qml.expval(qml.PauliZ(wires = 1)@qml.Identity(wires = 0))
    return qml.expval(qml.Identity(wires=0)@qml.PauliZ(wires=1))
```

### Actual behavior

Actual Behaviour:
`(qml.PauliZ(1)@qml.Identity(0)).matrix == (qml.PauliZ(0)@qml.Identity(1)).matrix ` is true
but
`(qml.Identity(0)@qml.PauliZ(1)).matrix == (qml.PauliZ(1)@qml.Identity(0)).matrix ` is false

### System information

```shell
Name: PennyLane
Version: 0.21.0
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/XanaduAI/pennylane
Author: 
Author-email: 
License: Apache License 2.0
Location: /Users/patrickhumbeli/anaconda3/envs/JAX/lib/python3.10/site-packages
Requires: appdirs, autograd, autoray, cachetools, networkx, numpy, pennylane-lightning, retworkx, scipy, semantic-version, toml
Required-by: PennyLane-Lightning
Platform info:           macOS-12.1-x86_64-i386-64bit
Python version:          3.10.2
Numpy version:           1.21.5
Scipy version:           1.8.0
Installed devices:
- lightning.qubit (PennyLane-Lightning-0.21.0)
- default.gaussian (PennyLane-0.21.0)
- default.mixed (PennyLane-0.21.0)
- default.qubit (PennyLane-0.21.0)
- default.qubit.autograd (PennyLane-0.21.0)
- default.qubit.jax (PennyLane-0.21.0)
- default.qubit.tf (PennyLane-0.21.0)
- default.qubit.torch (PennyLane-0.21.0)
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

**Comment by antalszava** [comment]
Hi @PatrickHuembeli, thank you for reporting this! :slightly_smiling_face: 

1.
> (qml.PauliZ(1)@qml.Identity(0)).matrix == (qml.PauliZ(0)@qml.Identity(1)).matrix is false

This is intended because we don't make any assumptions on the wire order. The reason is that we may have custom wire labels, e.g.
`(qml.PauliZ('a')@qml.Identity(0)).matrix == (qml.PauliZ(0)@qml.Identity('a')).matrix is false`.

A new way of obtaining operator matrices using the `qml.matrix` top-level function has just made it into our `master` branch (see https://github.com/PennyLaneAI/pennylane/pull/2241). This function will then accept a `wire_order` argument that should help get the expected results. :slightly_smiling_face: 

2.
As for the code snippet, thanks for reporting it, that should be working! We'll have a look into this.

**Comment by josh146** [comment]
Regarding point (2), I think the issue might be these lines here, inside the PennyLane `QubitDevice` class:

https://github.com/PennyLaneAI/pennylane/blob/b494a2bf1b68156a9688178e1d97dc348ebf1583/pennylane/_qubit_device.py#L782-L783

Here, the eigenvalue order is being computed by _sorting the wires_:

```pycon
>>> op1 = qml.PauliZ(1) @ qml.Identity(0)
>>> qml.eigvals(op1)
array([ 1, -1,  1, -1])
>>> np.kron([1, 1], [1, -1])
array([ 1, -1,  1, -1])
>>> op2 = qml.Identity(0) @ qml.PauliZ(1)
>>> qml.eigvals(op2)
array([ 1, -1,  1, -1])
```

So the eigenvalues of `op1` and `op2` are both being computed by assuming wire 0 < wire 1.

The problem is that the probability computation is not doing this; it is preserving the order of the observables in the tensor product, as it is being called with `op.wires`:

```pycon
>>> op1.wires
<Wires = [1, 0]>
>>> op2.wires
<Wires = [0, 1]>
```

So there is a mismatch between how a tensor product observable computes its matrix, and how it reports its wires. The question then becomes: which is correct? As long as we ensure both have the same convention, this will be fixed.

**Comment by josh146** [comment]
If we look at the `Tensor` product itself, the eigenvalues are being sorted here:

https://github.com/PennyLaneAI/pennylane/blob/f60407fc66ff2965c3d041a66cd405d17cb927a4/pennylane/operation.py#L1845-L1850

If we _remove this sorting_, this fixes this bug!

**Comment by Jaybsoni** [comment]
Hey @josh146, 

That is exactly correct!

I think if we remove the sort in the eigenvalues then we will be taking the convention in which the wire labels are ignored. While this makes sense in the case where custom wire labels are used (i.e wires = ["a", "b", "c"]), it is less intuitive than using the sorted order as the wires provided are ignored. I think in the majority of cases, taking the sorted convention makes more sense/.  

> Regarding point (2), I think the issue might be these lines here, inside the PennyLane `QubitDevice` class:
> 
> https://github.com/PennyLaneAI/pennylane/blob/b494a2bf1b68156a9688178e1d97dc348ebf1583/pennylane/_qubit_device.py#L782-L783
> 
> Here, the eigenvalue order is being computed by _sorting the wires_:
> 
> ```
> >>> op1 = qml.PauliZ(1) @ qml.Identity(0)
> >>> qml.eigvals(op1)
> array([ 1, -1,  1, -1])
> >>> np.kron([1, 1], [1, -1])
> array([ 1, -1,  1, -1])
> >>> op2 = qml.Identity(0) @ qml.PauliZ(1)
> >>> qml.eigvals(op2)
> array([ 1, -1,  1, -1])
> ```
> 
> So the eigenvalues of `op1` and `op2` are both being computed by assuming wire 0 < wire 1.
> 
> The problem is that the probability computation is not doing this; it is preserving the order of the observables in the tensor product, as it is being called with `op.wires`:
> 
> ```
> >>> op1.wires
> <Wires = [1, 0]>
> >>> op2.wires
> <Wires = [0, 1]>
> ```
> 
> So there is a mismatch between how a tensor product observable computes its matrix, and how it reports its wires. The question then becomes: which is correct? As long as we ensure both have the same convention, this will be fixed.

**Comment by josh146** [comment]
> I think if we remove the sort in the eigenvalues then we will be taking the convention in which the wire labels are ignored. While this makes sense in the case where custom wire labels are used, it is less intuitive than using the sorted order as the wires provided are ignored. I think in the majority of cases, taking the sorted convention makes more sense/.

@Jaybsoni something to keep in mind is that we shouldn't treat integer wires differently from other wire labels, as this introduces more inconsistency! 

Say we consider the case

```python
op1 = qml.PauliZ(1) @ qml.Identity(0)
op2 = qml.PauliZ(0) @ qml.Identity(1)
op3 = qml.PauliZ("a") @ qml.Identity(7)
```

Since we have no context as to what the canonical wire ordering _is_, I would argue that it doesn't make sense to attempt to order the eigenvalues by default! Instead, by default, we should expect all three of the above to return eigenvalues in the same order:

```python
>>> np.kron([1, -1], [1, 1])
array([ 1,  1, -1, -1])
```

However, this is not the case, since `Tensor.eigvals` is doing an implicit sort by default:

```python
>>>> qml.eigvals(op3)
array([ 1, -1,  1, -1])
>>> qml.eigvals(op2)
array([ 1,  1, -1, -1])
>>> qml.eigvals(op1)
array([ 1, -1,  1, -1])
```

It is making an assumption that may have no basis.

Instead, we should do the following here:

- If no canonical wire order is known, `qml.eigvals()` should make no assumptions, and simply use the ordering of the tensor product observables.

- If a canonical wire ordering _is_ known, we should be able to pass it to `qml.eigvals()` to take into account.

This is identical to how `qml.matrix()` behaves, and we should bring `qml.eigvals()` inline with this behaviour:

```python
>>> np.all(qml.matrix(op1) == qml.matrix(op2))
True
>>> np.all(qml.matrix(op1) == qml.matrix(op3))
True
>>> np.all(qml.matrix(op1, wire_order=[0, 1]) == qml.matrix(op2, wire_order=[0, 1]))
False
```

**Comment by josh146** [comment]
So this is partly why I lean towards _not_ sorting eigvals by default, while leaving the permutation in `probs` --- it seems a lot cleaner and consistent

**Comment by antalszava** [comment]
Hi @josh146,

> The problem is that the probability computation is not doing this; it is preserving the order of the observables in the tensor product

Isn't this being done here?
https://github.com/PennyLaneAI/pennylane/blob/67205b6231abb1380314597858cd3759c79a21a0/pennylane/_qubit_device.py#L757

This might even be the cause of the issue and avoiding this reordering might be a start on a viable solution. If no reordering is done, then the amplitudes remain in a big-endian ordering. It should then suffice to simply just to sort the observables (something that's already done). The reason why this sorting will be important is that as per the original example in 2., a user might provide the same observable in different forms and we'd like the same expectation to be calculated. Since the order of the tensor product terms (and hence the eigenvalues) is what may change in the user code, logically, it would make sense to change that only.

Attempting an even more complex example like `qml.Identity(wires=0) @ qml.Identity(wires=1) @ qml.PauliZ(wires=2)`

#2292 attempts this approach (having started first with changing the eigvals logic) with a few caveats:
* Tests that use `qml.Hermitian` fail: these cases depended on the reordering of amplitudes;
* Omitting the reordering in `marginal_prob` would likely require the intro of some new option or API: the `qml.probs` pipeline would still require this option to be there, whereas `qml.expval` may not.

**Comment by josh146** [comment]
@antalszava I see your point 🤔 My worry is that sorting of eigenvalues is a lot more difficult, both conceptually and logically, compared to sorting of probabilities.

E.g., by using the following diff, consider the output:

<details>
  <summary>Click to view PennyLane diff</summary>
  
```diff
diff --git a/pennylane/_qubit_device.py b/pennylane/_qubit_device.py
index 450f56210..06b0fb624 100644
--- a/pennylane/_qubit_device.py
+++ b/pennylane/_qubit_device.py
@@ -736,6 +736,7 @@ class QubitDevice(Device):
 
         # translate to wire labels used by device
         device_wires = self.map_wires(wires)
+        print("\nProbability device wires:", device_wires)
         inactive_device_wires = self.map_wires(inactive_wires)
 
         # reshape the probability so that each axis corresponds to a wire
@@ -780,6 +781,8 @@ class QubitDevice(Device):
                 ) from e
 
             prob = self.probability(wires=observable.wires)
+            print("eigvals:", eigvals)
+            print("probs:", prob)
             return self._dot(eigvals, prob)
 
         # estimate the ev
diff --git a/pennylane/operation.py b/pennylane/operation.py
index c27b88d63..f92c650c1 100644
--- a/pennylane/operation.py
+++ b/pennylane/operation.py
@@ -1845,19 +1845,12 @@ class Tensor(Observable):
         # observable should be Z^{\otimes n}
         self._eigvals_cache = pauli_eigs(len(self.wires))
 
-        # Sort observables lexicographically by the strings of the wire labels
-        # TODO: check for edge cases of the sorting, e.g. Tensor(Hermitian(obs, wires=[0, 2]),
-        # Hermitian(obs, wires=[1, 3, 4])
-        # Sorting the observables based on wires, so that the order of
-        # the eigenvalues is correct
-        obs_sorted = sorted(self.obs, key=lambda x: [str(l) for l in x.wires.labels])
-
         # check if there are any non-standard observables (such as Identity)
         if set(self.name) - standard_observables:
             # Tensor product of observables contains a mixture
             # of standard and non-standard observables
             self._eigvals_cache = np.array([1])
-            for k, g in itertools.groupby(obs_sorted, lambda x: x.name in standard_observables):
+            for k, g in itertools.groupby(self.obs, lambda x: x.name in standard_observables):
                 if k:
                     # Subgroup g contains only standard observables.
                     self._eigvals_cache = np.kron(self._eigvals_cache, pauli_eigs(len(list(g)))
```
</details>

```python
dev = qml.device("default.qubit", wires=2)
x = np.array(0.543, requires_grad=True)

op1 = qml.PauliZ(1) @ qml.Identity(0)
op2 = qml.Identity(0) @ qml.PauliZ(1)

@qml.qnode(dev)
def circuit(x, op):
    qml.RX(x, wires=0)
    qml.RY(x, wires=1)
    qml.CNOT(wires=[0, 1])
    return qml.expval(op)
```
```pycon
>>> print(circuit(x, op1))
Probability device wires: [1, 0]
eigvals: [ 1.  1. -1. -1.]
probs: [0.86133473 0.00517231 0.06674648 0.06674648]
0.7330140825988753

>>> print(circuit(x, op2))
Probability device wires: [0, 1]
eigvals: [ 1. -1.  1. -1.]
probs: [0.86133473 0.06674648 0.00517231 0.06674648]
0.7330140825988753

>>> print(circuit(x, qml.PauliZ(1)))
Probability device wires: [1]
eigvals: [ 1. -1.]
probs: [0.86650704 0.13349296]
0.7330140825988753
```

Having said that, your point is very valid --- it is fair to either remove the eigvals sorting, _or_ remove the probs sorting, both would solve this issue.

**Note**: If we do keep the `eigvals` sorting, we should require passing a `wire_order` argument,

```pycon
>>> op.get_eigvals(wire_order=...)
```

If `wire_order=None`, then no eigenvalue sorting is performed. This ensures that:

- `eigvals` behaves identically to other representations, like matrices
- we no longer treat integer wires differently to other wire labels.

Once this is done, we then have a choice inside the simulator:

1. Choice one: eigenvalue sorting, no probs sorting

   ```python
   eigvals = self._asarray(observable.get_eigvals(wire_order=self.wires), dtype=self.R_DTYPE)
   prob = self.probability()
   ```

2. Choice two: no eigenvalue sorting, probability sorting

   ```python
   eigvals = self._asarray(observable.get_eigvals(), dtype=self.R_DTYPE)
   prob = self.probability(observable.wires)
   ```

My intuition is that (2) will be easier to implement than (1).

**Comment by antalszava** [comment]
Hi @josh146,

> My worry is that sorting of eigenvalues is a lot more difficult, both conceptually and logically, compared to sorting of probabilities.

Agreed! However, it seems this would be the way that we eventually get the correct results in more intricate cases too.

While removing the observable sorting does help with the original 2-term tensor product case, it [fails when we have 3-terms in the TP](https://github.com/PennyLaneAI/pennylane/runs/5481305005?check_suite_focus=true#step:10:38). See the `test_pauliz_identity` device test and the https://github.com/PennyLaneAI/pennylane/pull/2293 prototype.

The same test passes on https://github.com/PennyLaneAI/pennylane/pull/2292.

**Comment by Jaybsoni** [comment]
This issue has been resolved in this [PR](https://github.com/PennyLaneAI/pennylane/pull/2276)
